### ISSUE TITLE
Rid ghc of (literate) perl?

### ISSUE THREAD

**Issue by angerman** [snowleopard/hadrian#139]
Tasks:
- [x] rewrite `driver/split/ghc-split.lprl` in haskell
  - This has already been done by @Mistuke, see https://ghc.haskell.org/trac/ghc/ticket/9832
- [ ] drop `Unlit` from the build system
- [ ] drop `Perl` from the build system
- [x] Optinally rewrite `utils/count_lines/count_lines.lprl` in haskell
- [ ] push changes upstream into `ghc` through phabricator
## 

ghc has two literate perl files:

```
./driver/split/ghc-split.lprl
./utils/count_lines/count_lines.lprl
```

as well as one regular perl file:

```
./nofib/runstdtest/runstdtest.prl
```

There are also a bunch of `.pl` files:

```
./distrib/remilestoning.pl
./libraries/containers/benchmarks/bench-cmp.pl
./libraries/hpc/tests/hpcrun.pl
./libraries/integer-gmp/gmp/gmpbuild/demos/perl/sample.pl
./libraries/integer-gmp/gmp/gmpbuild/demos/perl/test.pl
./libraries/integer-gmp/gmp/gmpbuild/demos/perl/test2.pl
./libraries/integer-gmp/gmp/gmpbuild/mpn/alpha/ev6/slot.pl
./libraries/integer-gmp/gmp/gmpbuild/mpn/m68k/t-m68k-defs.pl
./libraries/integer-gmp/gmp/gmpbuild/mpn/x86/k6/cross.pl
./libraries/integer-gmp/gmp/gmpbuild/mpn/x86/t-zdisp2.pl
./libraries/integer-gmp/gmp/gmpbuild/tune/many.pl
./nofib/fibon/Hackage/Cpsa/src/cpsa.pl
./nofib/fibon/Hackage/Cpsa/src/pp.pl
./nofib/fibon/Hackage/Cpsa/src/sexpr.pl
./utils/genargs/genargs.pl
```

of which none is part of the build system.

I can also find `build-perl` only used for the first two.

```
driver/split/ghc.mk
19:$(eval $(call build-perl,driver/split,dist))

rules/build-perl.mk
19:# $(eval $(call build-perl,driver/mangler,dist))

utils/count_lines/ghc.mk
5:$(eval $(call build-perl,utils/count_lines,dust))
```

`driver/mangler` seems to have been removed in

```
commit 9be618cdf99b04ce7eef6eeabc168b59174bb843
Author: Ian Lynagh <igloo@earth.li>
Date:   Mon Apr 4 21:59:04 2011 +0100

    Remove dead code, now that -fvia-c is a no-op
```

The use of `lprl`, adds quite a bit of complexity to the build system (#138), and requires the external tool `perl` (#83), and building the `Unlit` tool (#82).

**Comment by ndmitchell** [comment]
I would one day like to get rid of all the perl and much of the shell script, folding a lot into the Haskell build system. Note boot is perl, without any extension.

**Comment by angerman** [comment]
@ndmitchell yes, absolutely! The amount of extra complexity (wrappers, build-perl) we added to the build system is really sad.  I guess it's there for history reasons, and because make has it's limits as well.

But getting rid of `count_lines.lprl` (which isn't even really `literal` to begin with; it only has `\begin{code}` and `\end{code}`.) and `ghc-split`, which then later could be rolled into a library and used from `ghc` directly, without shelling out to a separate process, would be a not so bad start, I guess.

**Comment by ndmitchell** [comment]
Yep, seems sensible to me. Although may require changes to real GHC, and thus beyond this repo?

**Comment by ndmitchell** [comment]
Perhaps commit non literate versions here for now, and then upstream to GHC later? An upstream dir in this repo?

**Comment by angerman** [comment]
Yep. That would be the idea, replace `ghc-split` and `count_lines` by haskell packages here, and then initiate the upstream if we know it works for us :)

**Comment by snowleopard** [comment]
Sounds good to me! I'd be happy if we could get rid of perl bits. By the way, so far we didn't need to use `count_lines`, so maybe we won't need it at all?

**Comment by angerman** [comment]
Updated the comment to add the subtasks.

**Comment by angerman** [comment]
Ohh! @ndmitchell unearthed the following related ghc ticket on Twitter
https://ghc.haskell.org/trac/ghc/ticket/9832

cf: https://twitter.com/ndm_haskell/status/686101954301440000

Maybe we should just let this be and wait for 8.2?

**Comment by snowleopard** [comment]
I think it may be reasonable to wait. That is if this indeed happens in 8.2.

**Comment by angerman** [comment]
My understanding is that `-split-objs` is going to be replaced by `-split-sections`, which is currently blocked on `--gc-sections` support in `binutils`. It does add quite a bit of complexity and dependencies to the build system. In any case the whole idea behind `-split-objs` is to get away with a smaller resulting product. Now, maybe we can get away with completely dropping the `-split-objs` support from `shaking-up-ghc`? That would basically mean drop all the code I just wrote and some more, but I'd be fine with that.

Maybe @Mistuke and @bgamari can shed some additional insight?

**Comment by Mistuke** [comment]
I'm not really sure what the plan here is. My guess is that it is planned to drop `-split-objs` entirely in favor of `-split-sections` at some point. But I don't know when. `-split-sections` support particularly on Windows is a bit unknown at this point.

**Comment by bgamari** [comment]
D3567 rewrites `boot` in Python, removing that particular need.

**Comment by izgzhen** [comment]
I did something related here: https://github.com/snowleopard/hadrian/pull/312/commits/f3fa6d45fa77135fdd48f9793ee4ebee8c0aaa87

**Comment by snowleopard** [comment]
@izgzhen Nice! If this passes CI on all platforms, shall we put this into a separate PR?

**Comment by snowleopard** [comment]
Ah, it's not tested by the CI, because it requires `--split-objects` command line flag. Have you tried if it works for you?

**Comment by izgzhen** [comment]
@snowleopard I didn't see major outage, but I will keep an eye on this. I agree that it is useful to separate it out as another PR, since the `install-rule` one, as I expect, would take quite some time to resolve cleanly, and I can also make better sense of my change of dropping `Perl`.

**Comment by snowleopard** [comment]
This has partly been resolved: we no longer have any literate Perl in GHC and the corresponding build rule has been removed from Hadrian. 

As for removing Perl from GHC entirely, this seems to be more suited as a GHC ticket, not a Hadrian one. And there is, for example, https://ghc.haskell.org/trac/ghc/ticket/9832.

So, I think we can close this issue. When there is no Perl in GHC, dropping the Perl builder from Hadrian will be trivial. Please feel free to reopen if I am missing something.
